File: sdk-go/internal/testcommon/assert.go
package testcommon

import (
	"bytes"
	"encoding/json"
	"fmt"
	"regexp"
	"testing"

	llmsdk "github.com/hoangvvo/llm-sdk/sdk-go"
)

// TextPartAssertion represents an assertion for text parts
type TextPartAssertion struct {
	Text     *regexp.Regexp
	Citation *CitationAssertion
}

type CitationAssertion struct {
	Source    *regexp.Regexp
	Title     *regexp.Regexp
	CitedText *regexp.Regexp
}

func (a CitationAssertion) matches(citation llmsdk.Citation) bool {
	if a.Source != nil && !a.Source.MatchString(citation.Source) {
		return false
	}
	if a.Title != nil && (citation.Title == nil || !a.Title.MatchString(*citation.Title)) {
		return false
	}
	if a.CitedText != nil && (citation.CitedText == nil || !a.CitedText.MatchString(*citation.CitedText)) {
		return false
	}
	return true
}

func (a TextPartAssertion) matchesCitation(citations []llmsdk.Citation) bool {
	if a.Citation == nil {
		return true
	}
	for _, citation := range citations {
		if a.Citation.matches(citation) {
			return true
		}
	}
	return false
}

func (a TextPartAssertion) Assert(t *testing.T, content []llmsdk.Part) {
	t.Helper()

	for _, part := range content {
		if part.TextPart != nil && a.Text.MatchString(part.TextPart.Text) && a.matchesCitation(part.TextPart.Citations) {
			return
		}
	}

	contentJSON, _ := json.MarshalIndent(content, "", "  ")
	t.Errorf("Expected matching text part:\nExpected: %s\nReceived:\n%s", a.Text.String(), string(contentJSON))
}

// ToolCallPartAssertion represents an assertion for tool call parts
type ToolCallPartAssertion struct {
	ToolName string
	Args     *regexp.Regexp
}

// matchToolCallArgs matches tool call arguments against assertions
func matchToolCallArgs(raw json.RawMessage, expected *regexp.Regexp) bool {
	if expected == nil {
		return true
	}

	var compact bytes.Buffer
	if err := json.Compact(&compact, raw); err != nil {
		return false
	}

	return expected.Match(compact.Bytes())
}

func (a ToolCallPartAssertion) Assert(t *testing.T, content []llmsdk.Part) {
	t.Helper()

	for _, part := range content {
		if part.ToolCallPart != nil &&
			part.ToolCallPart.ToolName == a.ToolName &&
			matchToolCallArgs(part.ToolCallPart.Args, a.Args) {
			return
		}
	}

	contentJSON, _ := json.MarshalIndent(content, "", "  ")
	t.Errorf("Expected matching tool call part:\nExpected tool %s with args %+v\nReceived:\n%s", a.ToolName, a.Args, string(contentJSON))
}

type AudioPartAssertion struct {
	AudioID    bool // whether to check for audio ID presence
	Transcript *regexp.Regexp
}

func (a AudioPartAssertion) Assert(t *testing.T, content []llmsdk.Part) {
	t.Helper()

	for _, part := range content {
		if part.AudioPart != nil {
			if len(part.AudioPart.Data) == 0 {
				t.Errorf("Audio data must be present")
				return
			}
			if a.AudioID && (part.AudioPart.ID == nil || *part.AudioPart.ID == "") {
				t.Errorf("Expected audio ID to be present")
				return
			}
			if a.Transcript != nil {
				if part.AudioPart.Transcript == nil || !a.Transcript.MatchString(*part.AudioPart.Transcript) {
					t.Errorf("Expected matching transcript:\nExpected: %s\nReceived: %v", a.Transcript.String(),
						part.AudioPart.Transcript)
					return
				}
			}
			// pass
			return
		}
	}

	contentJSON, _ := json.MarshalIndent(content, "", "  ")
	t.Errorf("Expected matching audio part:\nReceived:\n%s", string(contentJSON))
}

type ImagePartAssertion struct {
	ImageID bool // whether to check for image ID presence
}

func (a ImagePartAssertion) Assert(t *testing.T, content []llmsdk.Part) {
	t.Helper()

	for _, part := range content {
		if part.ImagePart != nil {
			if len(part.ImagePart.Data) == 0 {
				t.Errorf("Image data must be present")
				return
			}
			if a.ImageID && (part.ImagePart.ID == nil || *part.ImagePart.ID == "") {
				t.Errorf("Expected image ID to be present")
				return
			}

			// pass
			return
		}
	}

	contentJSON, _ := json.MarshalIndent(content, "", "  ")
	t.Errorf("Expected matching image part:\nReceived:\n%s", string(contentJSON))
}

type ReasoningPartAssertion struct {
	Text      *regexp.Regexp
	Signature bool
}

func (a ReasoningPartAssertion) matches(part *llmsdk.ReasoningPart) bool {
	textMatches := a.Text.MatchString(part.Text)
	signatureMatches := !a.Signature || (part.Signature != nil && *part.Signature != "")
	return textMatches && signatureMatches
}

func (a ReasoningPartAssertion) Assert(t *testing.T, content []llmsdk.Part) {
	t.Helper()

	for _, part := range content {
		if part.ReasoningPart != nil && a.matches(part.ReasoningPart) {
			return
		}
	}

	contentJSON, _ := json.MarshalIndent(content, "", "  ")
	t.Errorf("Expected matching reasoning part:\nExpected: %s\nReceived:\n%s", a.Text.String(), string(contentJSON))
}

// PartAssertion represents assertions for different part types
type PartAssertion struct {
	TextPart      *TextPartAssertion
	ToolCallPart  *ToolCallPartAssertion
	ReasoningPart *ReasoningPartAssertion
	AudioPart     *AudioPartAssertion
	ImagePart     *ImagePartAssertion
}

// TestMethod represents the test method type
type TestMethod int

const (
	Generate TestMethod = iota
	Stream
)

// OutputAssertion represents the expected output
type OutputAssertion struct {
	Content []PartAssertion
}

// RunTests executes multiple test cases
func RunTests(t *testing.T, testCases []TestCase, model llmsdk.LanguageModel) {
	for _, testCase := range testCases {
		t.Run(fmt.Sprintf("%s / %s", model.Provider(), testCase.Name), func(t *testing.T) {
			RunTestCase(t, model, testCase.Name)
		})
	}
}

// assertContentPart asserts that content matches assertions
func assertContentPart(t *testing.T, content []llmsdk.Part, assertions []PartAssertion) {
	t.Helper()

	for _, assertion := range assertions {
		if assertion.TextPart != nil {
			assertion.TextPart.Assert(t, content)
		}
		if assertion.ToolCallPart != nil {
			assertion.ToolCallPart.Assert(t, content)
		}
		if assertion.AudioPart != nil {
			assertion.AudioPart.Assert(t, content)
		}
		if assertion.ImagePart != nil {
			assertion.ImagePart.Assert(t, content)
		}
		if assertion.ReasoningPart != nil {
			assertion.ReasoningPart.Assert(t, content)
		}
	}
}

// NewTextAssertion creates a new text part assertion
func NewTextAssertion(pattern string) PartAssertion {
	return PartAssertion{
		TextPart: &TextPartAssertion{
			Text: regexp.MustCompile("(?s)" + pattern),
		},
	}
}

func NewTextAssertionWithCitation(pattern string, citation CitationAssertion) PartAssertion {
	assertion := NewTextAssertion(pattern)
	assertion.TextPart.Citation = &citation
	return assertion
}

// NewToolCallAssertion creates a new tool call part assertion
func NewToolCallAssertion(toolName string, argsPattern string) PartAssertion {
	var args *regexp.Regexp
	if argsPattern != "" {
		args = regexp.MustCompile("(?s)" + argsPattern)
	}
	return PartAssertion{
		ToolCallPart: &ToolCallPartAssertion{
			ToolName: toolName,
			Args:     args,
		},
	}
}

func NewReasoningAssertion(textPattern string, signature ...bool) PartAssertion {
	hasSignature := len(signature) > 0 && signature[0]
	return PartAssertion{
		ReasoningPart: &ReasoningPartAssertion{
			Text:      regexp.MustCompile("(?s)" + textPattern),
			Signature: hasSignature,
		},
	}
}

func NewAudioAssertion(audioID bool, transcriptPattern string) PartAssertion {
	var transcriptRegex *regexp.Regexp
	if transcriptPattern != "" {
		transcriptRegex = regexp.MustCompile("(?s)" + transcriptPattern)
	}
	return PartAssertion{
		AudioPart: &AudioPartAssertion{
			AudioID:    audioID,
			Transcript: transcriptRegex,
		},
	}
}

func NewImageAssertion(imageID bool) PartAssertion {
	return PartAssertion{
		ImagePart: &ImagePartAssertion{
			ImageID: imageID,
		},
	}
}

File: sdk-js/test-common/assert.ts
import { type TestContext } from "node:test";
import type { Part } from "../src/types.ts";

interface TextPartAssertion {
  type: "text";
  text: RegExp;
}

interface ToolCallPartAssertionArgProp {
  [key: string]: RegExp | ToolCallPartAssertionArgProp;
}

export type PartAssertion =
  | TextPartAssertion
  | ToolCallPartAssertion
  | ReasoningPartAssertion
  | AudioPartAssertion;

export interface ToolCallPartAssertion {
  type: "tool_call";
  tool_name: string;
  args: ToolCallPartAssertionArgProp;
}

export interface AudioPartAssertion {
  type: "audio";
  id?: boolean;
  transcript?: RegExp | undefined;
}

export interface ReasoningPartAssertion {
  type: "reasoning";
  text: RegExp;
}

export function assertContentPart(
  t: TestContext,
  content: Part[],
  assertions: PartAssertion[],
) {
  for (const assertion of assertions) {
    switch (assertion.type) {
      case "text": {
        assertTextPart(t, content, assertion);
        break;
      }
      case "tool_call": {
        assertToolCallPart(t, content, assertion);
        break;
      }
      case "audio": {
        assertAudioPart(t, content, assertion);
        break;
      }
      case "reasoning": {
        assertReasoningPart(t, content, assertion);
        break;
      }
    }
  }
}

export function assertTextPart(
  t: TestContext,
  content: Part[],
  assertion: TextPartAssertion,
) {
  const foundPart = content.find(
    (part) => part.type === "text" && assertion.text.test(part.text),
  );
  t.assert.ok(
    foundPart,
    `Expected matching text part:
Expected: ${String(assertion.text)}
Received:
${JSON.stringify(content, null, 2)}`,
  );
}

export function assertToolCallPart(
  t: TestContext,
  content: Part[],
  assertion: ToolCallPartAssertion,
) {
  const foundPart = content.find(
    (part) =>
      part.type === "tool-call" &&
      part.tool_name === assertion.tool_name &&
      matchToolCallArgs(part.args, assertion.args),
  );
  t.assert.ok(
    foundPart,
    `Expected matching tool call part:
Expected tool ${assertion.tool_name} with args ${JSON.stringify(assertion.args)}
Received:
${JSON.stringify(content, null, 2)}`,
  );
}

export function assertAudioPart(
  t: TestContext,
  content: Part[],
  assertion: AudioPartAssertion,
) {
  const foundPart = content.find((part) => {
    if (part.type !== "audio") {
      return false;
    }
    if (!part.data) {
      return false;
    }
    if (assertion.id && !part.id) {
      return false;
    }
    if (assertion.transcript && !assertion.transcript.test(part.transcript!)) {
      return false;
    }
    return true;
  });
  t.assert.ok(
    foundPart,
    `Expected matching audio part:
Expected: ${JSON.stringify(assertion)}
Received:
${JSON.stringify(content, null, 2)}`,
  );
}

export function assertReasoningPart(
  t: TestContext,
  content: Part[],
  assertion: ReasoningPartAssertion,
) {
  const foundPart = content.find(
    (part) => part.type === "reasoning" && assertion.text.test(part.text),
  );
  t.assert.ok(
    foundPart,
    `Expected matching reasoning part:
Expected text ${String(assertion.text)}
Received:
${JSON.stringify(content, null, 2)}`,
  );
}

function matchToolCallArgs(
  actual: Record<string, unknown>,
  expected: ToolCallPartAssertionArgProp,
): boolean {
  for (const expectedKey in expected) {
    const expectedValue = expected[expectedKey]!;
    const actualValue = actual[expectedKey];

    if (actualValue === undefined) {
      return false;
    }

    if (expectedValue instanceof RegExp) {
      if (
        !(
          typeof actualValue !== "object" &&
          expectedValue.test(String(actualValue))
        )
      ) {
        return false;
      }
    } else {
      if (
        !(
          typeof actualValue === "object" &&
          matchToolCallArgs(
            actualValue as Record<string, unknown>,
            expectedValue,
          )
        )
      ) {
        return false;
      }
    }
  }

  return true;
}
